pep-9999.rst

```diff
@@ -0,0 +1,675 @@
+example, the following snippet is legal::
+
+   def expects_str(x: str) -> None: ...
+   def expects_literal(x: Literal["foo"]) -> None: ...
+
+   # Legal: "foo" is inferred to be of type 'str'
+   expects_str("foo")
+
+   # Legal: "foo" is inferred to be of type 'Literal["foo"]'
+   expects_literal("foo")
+
+However, non-literal expressions in general will not automatically be inferred
+to be literals. For example::
+
+   def expects_literal(x: Literal["foo"]) -> None: ...
+
+   def runner(my_str: str) -> None:
+       # ILLEGAL: str is not a subclass of Literal["foo"]
+       expects_literal(my_str)
+
+**Note:** If the user wants their API to support accepting both literals
+*and* the original type -- perhaps for legacy purposes -- they should
+implement a fallback overload. See `Interactions with overloads`_.
+
+Miscellaneous interactions
+--------------------------
+
+This section discusses how literal types interact with other existing types.
+
+Intelligent indexing of structured data: Interactions with TypedDict, Tuple, NamedTuples, and getattr
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+Literals can be used to "intelligently index" into structured types
+TypedDicts, NamedTuple, and classes. (This is not an exhaustive list).
+
+For example, type checkers should infer the correct value type when
+indexing into a TypedDict using a string literal that corresponds to
+one of the available keys::
+
+   Foo = TypedDict('Foo', {
+       'key1': int,
+       'key2': str,
+   })
+
+   a: Final = "key1"
+   b: Final = "some other string"
+
+   f: Foo
+   reveal_type(f[a])  # Revealed type is 'int'
+   f[b]               # Error: 'Foo' does not contain a key named 'some other string'
+
+We require similar behavior when indexing into a tuple or NamedTuple::
+
+   a: Final = 0
+   b: Final = 5
+
+   some_tuple: Tuple[int, str, List[bool]] = (3, "abc", [True, False])
+   reveal_type(some_tuple[a])   # Revealed type is 'int'
+   some_tuple[b]                # Error: 5 is not a valid index into the tuple
+
+...and when using functions like getattr::
+
+   class Test:
+       def __init__(self, param: int) -> None:
+           self.myfield = param
+       
+       def mymethod(self, val: int) -> str: ...
+       
+   a: Final = "myfield"
+   b: Final = "mymethod"
+   c: Final = "blah"
+
+   t = Test()
+   reveal_type(getattr(t, a))  # Revealed type is 'int'
+   reveal_type(getattr(t, b))  # Revealed type is 'Callable[[int], str]'
+   getattr(t, c)               # Error: 'Test' does not have attribute named 'blah'
+
+These interactions will most likely need to be added to type checkers on
+an ad-hoc basis. This is a little unfortunate: it would have been nice to
+unify these interactions by adding  something like TypeScript’s
+`index types <typescript-index-types_>`_ and ``keyof`` operator, which lets
+you encode the idea that some key (e.g. a literal string) is a member of
+some object.
```

Comment:
Also deleted -- TBH I included this only because I got some feedback earlier to discuss existing work in TypeScript. This section was pretty shoehorned in, anyways.